tide/master/rest/RestWindows.cpp

```diff
@@ -43,33 +43,71 @@
 #include "scene/ContentWindow.h"
 #include "thumbnail/thumbnail.h"
 
+#include "log.h"
+
 #include <QBuffer>
+#include <QtConcurrent>
+#include <QFuture>
 #include <QJsonArray>
 #include <QJsonDocument>
 #include <QJsonObject>
+#include <QString>
+#include <QThreadPool>
+
+using namespace zeroeq;
 
 namespace
 {
 const QRegExp _regex = QRegExp("\\{|\\}");
 const QSize thumbnailSize{ 512, 512 };
+
+QString _getUuid( const ContentWindow& window )
+{
+   return window.getID().toString().replace( _regex, "" );
+}
 }
 
-RestWindows::RestWindows( zeroeq::http::Server& server,
-                          const DisplayGroup& displayGroup )
-    : _httpServer( server ), _displayGroup( displayGroup )
+RestWindows::RestWindows( const DisplayGroup& displayGroup )
+    : _displayGroup( displayGroup )
 {
     connect( &displayGroup, &DisplayGroup::contentWindowAdded,
-              this, &RestWindows::_registerThumbnailUrl );
+             this, &RestWindows::_cacheThumbnail );
+
     connect( &displayGroup, &DisplayGroup::contentWindowRemoved,
-              this, &RestWindows::_deregisterThumbnailUrl );
+             [this]( ContentWindowPtr window )
+    {
+        const auto fileName = window->getContent()->getURI();
+        if( window->getContent()->getURI().startsWith(QDir::tempPath() + "/" ))
+        {
+            QDir().remove( fileName );
+            put_flog( LOG_INFO, "Removing file: %s",
+                      fileName.toLocal8Bit().constData( ));
+        }
+        _thumbnailCache.remove( _getUuid( *window ));
+    });
 }
 
-std::string RestWindows::getTypeName() const
+std::future<http::Response> RestWindows::getWindowInfo( const std::string& path,
+        const std::string& )
 {
-    return "tide/windows";
+    if( path.empty( ))
+    {
+        http::Response response;
+        response.payload = _getWindowList();
+        return make_ready_future( response );
+    }
+
+    const auto endpoint = QString::fromStdString( path );
+    if( endpoint.endsWith( "/thumbnail" ))
+    {
+        const auto uuid = endpoint.split( "/" )[0];
```

Comment:
that is only correct if there is a single "/" in the request. Split first, and return _getThumbnail() only if there are two parts and the second one == "thumbnail"